Add volumes config to LXC driver

mikemccracken opened the pull request:
Allow lxc driver to accept bind mount config similarly to the docker
driver.

Includes some static sanity checks in Validate step

Signed-off-by: Michael McCracken <mikmccra@cisco.com>

mikemccracken commented:
The failing check looks like a Travis-only error. I was unable to reproduce it locally.
I'm going to close and reopen this PR in case it is transient.

dadgar commented on client/driver/lxc.go:
```diff
@@ -250,6 +270,13 @@ func (d *LxcDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
 		fmt.Sprintf("%s secrets none rw,bind,create=dir", ctx.TaskDir.SecretsDir),
 	}
+
+	for _, volDesc := range driverConfig.Volumes {
```
Please guard this in a similar fashion to how it is done in Docker: https://github.com/hashicorp/nomad/blob/master/client/driver/docker.go#L85
https://github.com/hashicorp/nomad/blob/master/client/driver/docker.go#L496
https://github.com/hashicorp/nomad/blob/master/client/driver/docker.go#L1003

dadgar commented on client/driver/lxc.go:
```diff
@@ -250,6 +270,13 @@ func (d *LxcDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
+
+	for _, volDesc := range driverConfig.Volumes {
+		// the format was checked in Validate()
+		paths := strings.Split(volDesc, ":")
```
Add tests and documentation

mikemccracken commented:
My apologies if overwriting the first commit made the review hard to go back to. Let me know if there's a workflow you prefer to follow for Nomad.

schmichael commented:
> My apologies if overwriting the first commit made the review hard to go back to. Let me know if there's a workflow you prefer to follow for Nomad. -- @mikemccracken 

Nope! That's fine!

schmichael commented on client/driver/lxc_test.go:
```diff
@@ -200,3 +201,98 @@ func TestLxcDriver_Open_Wait(t *testing.T) {
+		[]string{
+			":def",
+		},
+		[]string{
```
Tests are failing because our linter prefers the simplified version of this code. You can fix it with either `gofmt -s lxc_test.go` or just by using this:

```go
        brokenVolumeConfigs := [][]string{
                {
                        "foo:/var",
                },
                {
                        ":",
                },
                {
                        "abc:",
                },
                {
                        ":def",
                },
                {
                        "abc:def:ghi",
                },
        }
```

A bit of a silly code style quibble, but it keeps our code consistent and linters happy!

mikemccracken commented:
Hi, just checking if there's anything else you need me to do here.

Thanks!

dadgar commented on client/driver/lxc_test.go:
```diff
@@ -200,3 +201,98 @@ func TestLxcDriver_Open_Wait(t *testing.T) {
+
+}
+
+func TestLxcDriver_Start_NoVolumes(t *testing.T) {
```
It would be nice to get one that tests actually mounting a volume

dadgar commented:
@mikemccracken Everything looks great. I would just like to see a test actually mount a volume

mikemccracken commented:
Rebased onto master, and added an extra check:

The original changes to `TestLxcDriver_Start_Wait()` mounted /tmp as /mnt/tmp in the container and just checked that the directory's mount point is created in the container. I added code to write a file and check that the file's contents as seen from inside the container match.

dadgar commented:
@mikemccracken Really great PR! Thank you
